### PR TITLE
feat: implement get_intermediate_states

### PR DESCRIPTION
**Pull request by redeboer**
The `print_intermediate_states` [defined in the quickstart](https://pwa.readthedocs.io/projects/expertsystem/en/latest/usage/quickstart.html#1.3.-Find-solutions) has been moved to the `ui` module (slightly modified form).

### CODE REVIEW

File: examples/quickstart.ipynb [hunk 1]
```diff
@@ -56,10 +56,7 @@
    "metadata": {},
    "outputs": [],
    "source": [
-    "from expertsystem.ui import (\n",
-    "    InteractionTypes,\n",
-    "    StateTransitionManager,\n",
-    ")"
+    "from expertsystem.ui import InteractionTypes, StateTransitionManager"
```
Comment (spflueger) [inline examples/quickstart.ipynb:59]: why does this change, where the brackets removed or did black do this automatically?

File: expertsystem/ui/__init__.py [hunk 2]
```diff
@@ -459,3 +465,26 @@ def _particle_with_spin_projection_to_dict(
...
+def get_intermediate_state_names(
+    solutions: List[StateTransitionGraph],
+) -> Set[str]:
+    """Extract the names of the intermediate states in the solutions."""
+    intermediate_states = set()
+    for graph in solutions:
+        for edge_id in graph.get_intermediate_state_edges():
+            edge_property = graph.edge_props[edge_id]
+            if isinstance(edge_property, dict):
+                intermediate_states.add(edge_property["Name"])
```
Comment (spflueger) [inline expertsystem/ui/__init__.py:479]: we have to remember and clean up this one at the end. I would almost prefer if currently just the dict version is implemented (as thats whats currently returned by the solvers). Then once that is fixed its impossible to miss this change

File: examples/quickstart.ipynb (hunk 1 above)
Comment (redeboer) [reply examples/quickstart.ipynb:59]: Black formatting, apparently the line was too long. Not sure why it wasn't formatted before (haven't found a way to format notebooks from the terminal and integrate into the CI).

Comment (spflueger) [reply examples/quickstart.ipynb:59]: hmm i see

File: expertsystem/ui/__init__.py (hunk 2 above)
Comment (redeboer) [reply expertsystem/ui/__init__.py:479]: That's a good point. I'm working on moving this function to `state.properties`, but now I noticed that I messed up something in #269: the idea was to have that PR split `particles` into `nested_dicts`, `state.properties` and `state.combinatorics`... Follow-up soon

Comment (redeboer) [reply expertsystem/ui/__init__.py:479]: Btw, there are more of these 'backward compatible' functions, e.g. the ones introduced in #262.

Comment (spflueger) [reply expertsystem/ui/__init__.py:479]: Yes, if both cases are needed, then its of course unavoidable. I don't know the best way around this. My idea is to just spot those old unused cases via code coverage

**Review by spflueger — APPROVED** [review]

**Comment by redeboer** [comment]
Want to rebase this one on #272 and move `get_intermediate_state_names` to `state.properties`. Most of the remaining nested `dict` functionality is there (the bulk is of course under `nested_dict`).

**Comment by redeboer** [comment]
https://github.com/ComPWA/expertsystem/pull/270#issuecomment-698810019
Oh nvm, actually this function is too much of a facade to be placed into `state.properties`.

However, I still want #272 to pass first to keep the eventual squashed commit of that one as a follow-up to the one of #269.

**Pull request merged**